<a href="https://colab.research.google.com/github/dp2292/CP01-Data-Alchemists/blob/Devansh/CP01_Data_Alchemists.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Section 1: Data Pre-processing and Analysis

In [ ]:
!mkdir data

In [ ]:
!gdown --id 1RfhNYBd7g70LdgJROvx8iPGygPhiGQLZ -O data/

/usr/local/lib/python3.10/dist-packages/gdown/cli.py:121: FutureWarning: Option `--id` was deprecated in version 4.3.1 and will be removed in 5.0. You don't need to pass it anymore to use a file ID.
  warnings.warn(
Downloading...
From: https://drive.google.com/uc?id=1RfhNYBd7g70LdgJROvx8iPGygPhiGQLZ
To: /content/data/CP01.zip
100% 630M/630M [00:12<00:00, 51.8MB/s]


In [ ]:
!unzip /content/data/CP01.zip -d /content/data

Archive:  /content/data/CP01.zip
  inflating: /content/data/For_modeling.csv  
  inflating: /content/data/cleaned_seoul_bike_data.csv  


In [ ]:
import math
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
data = pd.read_csv('data/For_modeling.csv',index_col=0)
data = data.reset_index().drop(columns=['index'])
data.head(5)

Duration  Distance      PLong       PLatd      DLong       DLatd  \
0         3        50  37.544666  126.888359  37.544666  126.888359   
1        24      7670  37.506199  127.003944  37.551250  127.035103   
2         8      1390  37.544590  127.057083  37.537014  127.061096   
3         8      1820  37.571102  127.023560  37.561447  127.034920   
4         4       850  37.573242  127.015907  37.565849  127.016403   

   Haversine  Pmonth  Pday  Phour  ...  Dmin  DDweek  Temp  Precip  Wind  \
0   0.000000       1     1      0  ...     4       0  -3.2     0.0   0.5   
1   5.713529       1     1      0  ...    25       0  -3.2     0.0   0.5   
2   0.913702       1     1      0  ...     9       0  -3.2     0.0   0.5   
3   1.468027       1     1      0  ...    10       0  -3.2     0.0   0.5   
4   0.823227       1     1      0  ...     6       0  -3.2     0.0   0.5   

   Humid  Solar  Snow  GroundTemp  Dust  
0   40.0    0.0   0.0        -2.2  25.0  
1   40.0    0.0   0.0        -2.2  25.0  
2   40.0    0.0   0.0        -2.2  25.0  
3   40.0    0.0   0.0        -2.2  25.0  
4   40.0    0.0   0.0        -2.2  25.0  

[5 rows x 25 columns]

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9601139 entries, 0 to 9601138
Data columns (total 25 columns):
 #   Column      Dtype  
---  ------      -----  
 0   Duration    int64  
 1   Distance    int64  
 2   PLong       float64
 3   PLatd       float64
 4   DLong       float64
 5   DLatd       float64
 6   Haversine   float64
 7   Pmonth      int64  
 8   Pday        int64  
 9   Phour       int64  
 10  Pmin        int64  
 11  PDweek      int64  
 12  Dmonth      int64  
 13  Dday        int64  
 14  Dhour       int64  
 15  Dmin        int64  
 16  DDweek      int64  
 17  Temp        float64
 18  Precip      float64
 19  Wind        float64
 20  Humid       float64
 21  Solar       float64
 22  Snow        float64
 23  GroundTemp  float64
 24  Dust        float64
dtypes: float64(13), int64(12)
memory usage: 1.8 GB


In [ ]:
print(data.isna().sum().sort_values(ascending=False))

Duration      0
Dday          0
GroundTemp    0
Snow          0
Solar         0
Humid         0
Wind          0
Precip        0
Temp          0
DDweek        0
Dmin          0
Dhour         0
Dmonth        0
Distance      0
PDweek        0
Pmin          0
Phour         0
Pday          0
Pmonth        0
Haversine     0
DLatd         0
DLong         0
PLatd         0
PLong         0
Dust          0
dtype: int64


In [ ]:
data.describe()

Duration      Distance         PLong         PLatd         DLong  \
count  9.601139e+06  9.601139e+06  9.601139e+06  9.601139e+06  9.601139e+06   
mean   2.580294e+01  3.713519e+03  3.754761e+01  1.269908e+02  3.754770e+01   
std    2.504333e+01  3.956611e+03  4.441376e-02  8.257065e-02  4.445305e-02   
min    1.000000e+00  1.000000e+00  3.743727e+01  1.267986e+02  3.743727e+01   
25%    8.000000e+00  1.230000e+03  3.751424e+01  1.269200e+02  3.751424e+01   
50%    1.600000e+01  2.280000e+03  3.754693e+01  1.269943e+02  3.754655e+01   
75%    3.600000e+01  4.620000e+03  3.757324e+01  1.270618e+02  3.757324e+01   
max    1.190000e+02  3.329000e+04  3.768972e+01  1.271803e+02  3.768972e+01   

              DLatd     Haversine        Pmonth          Pday         Phour  \
count  9.601139e+06  9.601139e+06  9.601139e+06  9.601139e+06  9.601139e+06   
mean   1.269907e+02  1.870315e+00  7.578754e+00  1.576714e+01  1.437801e+01   
std    8.323726e-02  1.995361e+00  2.696797e+00  8.748012e+00  6.249244e+00   
min    1.267986e+02  0.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    1.269194e+02  6.758669e-01  6.000000e+00  8.000000e+00  1.000000e+01   
50%    1.269947e+02  1.257498e+00  8.000000e+00  1.600000e+01  1.600000e+01   
75%    1.270624e+02  2.363812e+00  1.000000e+01  2.300000e+01  1.900000e+01   
max    1.271803e+02  2.863448e+01  1.200000e+01  3.100000e+01  2.300000e+01   

       ...          Dmin        DDweek          Temp        Precip  \
count  ...  9.601139e+06  9.601139e+06  9.601139e+06  9.601139e+06   
mean   ...  2.999278e+01  2.945771e+00  1.856201e+01  1.784856e-02   
std    ...  1.730960e+01  1.977624e+00  9.498075e+00  3.736282e-01   
min    ...  0.000000e+00  0.000000e+00 -1.780000e+01  0.000000e+00   
25%    ...  1.500000e+01  1.000000e+00  1.220000e+01  0.000000e+00   
50%    ...  3.000000e+01  3.000000e+00  2.010000e+01  0.000000e+00   
75%    ...  4.500000e+01  5.000000e+00  2.550000e+01  0.000000e+00   
max    ...  5.900000e+01  6.000000e+00  3.940000e+01  3.500000e+01   

               Wind         Humid         Solar          Snow    GroundTemp  \
count  9.601139e+06  9.601139e+06  9.601139e+06  9.601139e+06  9.601139e+06   
mean   1.819274e+00  5.397565e+01  7.300679e-01  9.948924e-03  2.072376e+01   
std    9.710258e-01  1.742159e+01  9.175004e-01  1.628091e-01  1.265221e+01   
min    0.000000e+00  1.000000e+01  0.000000e+00  0.000000e+00 -1.360000e+01   
25%    1.100000e+00  4.100000e+01  0.000000e+00  0.000000e+00  1.160000e+01   
50%    1.700000e+00  5.300000e+01  2.700000e-01  0.000000e+00  2.110000e+01   
75%    2.400000e+00  6.700000e+01  1.260000e+00  0.000000e+00  2.850000e+01   
max    7.400000e+00  9.800000e+01  3.520000e+00  8.800000e+00  6.220000e+01   

               Dust  
count  9.601139e+06  
mean   3.245798e+01  
std    2.407264e+01  
min    0.000000e+00  
25%    1.600000e+01  
50%    2.700000e+01  
75%    4.200000e+01  
max    3.040000e+02  

[8 rows x 25 columns]

In [ ]:
data_with_zero_displacement = data[data['Haversine']==0]
data_with_zero_displacement.head(5)

Duration  Distance      PLong       PLatd      DLong       DLatd  \
0           3        50  37.544666  126.888359  37.544666  126.888359   
26         18      1600  37.556610  126.898018  37.556610  126.898018   
139         2        10  37.575970  126.983063  37.575970  126.983063   
159        43      1940  37.548222  127.067879  37.548222  127.067879   
206        13       490  37.609566  126.930977  37.609566  126.930977   

     Haversine  Pmonth  Pday  Phour  ...  Dmin  DDweek  Temp  Precip  Wind  \
0          0.0       1     1      0  ...     4       0  -3.2     0.0   0.5   
26         0.0       1     1      0  ...    27       0  -3.2     0.0   0.5   
139        0.0       1     1      0  ...    40       0  -3.2     0.0   0.5   
159        0.0       1     1      0  ...    26       0  -3.2     0.0   0.5   
206        0.0       1     1      0  ...     5       0  -3.2     0.0   0.5   

     Humid  Solar  Snow  GroundTemp  Dust  
0     40.0    0.0   0.0        -2.2  25.0  
26    40.0    0.0   0.0        -2.2  25.0  
139   40.0    0.0   0.0        -2.2  25.0  
159   40.0    0.0   0.0        -2.2  25.0  
206   40.0    0.0   0.0        -2.2  25.0  

[5 rows x 25 columns]

In [ ]:
data_with_zero_displacement.shape

(955083, 25)

In [ ]:
data.shape

(9601139, 25)

Out of 9601139 recorded trips, 955083 are round-trips, meaning the rider has dropped the bike from where he picked up it! Logically, it doesn't make sense to include this data in predicting future bike trips duration from one point to another as the pickup and dropoff point being same, we cannot get any idea about the rider's whereabouts in between the trip from this data.

In [ ]:
# Check if pickup and dropoff coordinates are the same
same_coordinates_mask = np.logical_and(data['PLong'] == data['DLong'], data['PLatd'] == data['DLatd'])

# Filter the data based on both conditions
filtered_data = data[~(same_coordinates_mask & (data['Haversine'] == 0))]

In [ ]:
filtered_data.shape

(8646056, 25)

In [ ]:
# removing the tuples for which haversine == 0
filtered_data = data[data['Haversine'] != 0]

In [ ]:
filtered_data.shape

(8646056, 25)